File: packages/ui-flutter/lib/src/virtual_list.dart
import 'package:flutter/widgets.dart';

/// Mirrors `packages/ui/src/components/virtual-list.tsx`: a windowed-rendering
/// primitive. Like `KinetixVirtualList` on Compose (`LazyColumn`) and SwiftUI
/// (`List`), this maps cleanly onto a native mechanism — `ListView.builder`
/// already only builds the children near the visible viewport, so there's no
/// manual scroll-offset math to port and no fixed-row-height restriction.
class KinetixVirtualList<T> extends StatelessWidget {
  const KinetixVirtualList({
    super.key,
    required this.items,
    required this.itemBuilder,
    this.itemExtent,
  });

  final List<T> items;
  final Widget Function(BuildContext context, T item, int index) itemBuilder;

  /// Fixed row height, forwarded to `ListView.builder`'s `itemExtent` when
  /// known — lets Flutter skip measuring every row up front.
  final double? itemExtent;

  @override
  Widget build(BuildContext context) {
    return ListView.builder(
      itemCount: items.length,
      itemExtent: itemExtent,
      itemBuilder: (context, index) => itemBuilder(context, items[index], index),
    );
  }
}

File: packages/ui-swiftui/Sources/KinetixUI/VirtualList.swift
//
// VirtualList.swift — KinetixVirtualList.
//
// Mirrors packages/ui/src/components/virtual-list.tsx: a windowed-rendering
// primitive. Like KinetixVirtualList on Compose (LazyColumn), this maps
// cleanly onto a native mechanism — SwiftUI's `List` already only
// instantiates the rows near the visible viewport, so there's no manual
// scrollTop/ResizeObserver math to port and no fixed-row-height
// restriction. `id` is an explicit key path (rather than requiring
// `T: Identifiable`) so this accepts the same shape of plain data the web
// and Compose versions do.
//

import SwiftUI

public struct KinetixVirtualList<Data: RandomAccessCollection, ID: Hashable, Content: View>: View {
    private let data: Data
    private let id: KeyPath<Data.Element, ID>
    private let content: (Data.Element) -> Content

    public init(
        _ data: Data,
        id: KeyPath<Data.Element, ID>,
        @ViewBuilder content: @escaping (Data.Element) -> Content
    ) {
        self.data = data
        self.id = id
        self.content = content
    }

    public var body: some View {
        List(data, id: id) { item in
            content(item)
                .listRowInsets(EdgeInsets())
                .listRowSeparator(.hidden)
        }
        .listStyle(.plain)
    }
}
